src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -33,10 +35,23 @@ object FunctionBuilder {
     new ClassReader(bytes).accept(tcv, 0)
   }
 
+  def functionBuilder[R](implicit rti: TypeInfo[R]): FunctionBuilder[AsmFunction0[R]] =
+    new FunctionBuilder(Array[MaybeGenericTypeInfo[_]](), GenericTypeInfo(rti), "is/hail/asm4s/AsmFunction0")
+
+  def functionBuilder[A,R](implicit ati: TypeInfo[A], rti: TypeInfo[R]): FunctionBuilder[AsmFunction1[A, R]] =
+    new FunctionBuilder(Array(GenericTypeInfo(ati)), GenericTypeInfo(rti), "is/hail/asm4s/AsmFunction1")
+
+  def functionBuilder[A,B,R](implicit ati: TypeInfo[A], bti: TypeInfo[B], rti: TypeInfo[R]): FunctionBuilder[AsmFunction2[A, B, R]] =
+    new FunctionBuilder(Array(GenericTypeInfo(ati), GenericTypeInfo(bti)), GenericTypeInfo(rti), "is/hail/asm4s/AsmFunction2")
+
+  private implicit def methodNodeToGrowable(mn: MethodNode): Growable[AbstractInsnNode] = new Growable[AbstractInsnNode] {
+    def +=(e: AbstractInsnNode) = { mn.instructions.add(e); this }
+    def clear() { throw new UnsupportedOperationException() }
+  }
 }
 
-abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnTypeInfo: TypeInfo[R],
-  packageName: String = "is/hail/codegen/generated") {
+class FunctionBuilder[F >: Null](parameterTypeInfo: Array[MaybeGenericTypeInfo[_]], returnTypeInfo: MaybeGenericTypeInfo[_],
+  interfaceName: String, packageName: String = "is/hail/codegen/generated") {
```

Comment:
One last question: can't you find out which parameters (or return values) are generic using reflection on F?  java.lang.reflect.Method has various functions for getting generic types, generic signatures, etc.  This PR is looking good, happy to leave this for the next round.